js/web/lib/wasm/jsep/webgpu/ops/scatter-nd.ts

```diff
@@ -0,0 +1,170 @@
+                  let res = atomicCompareExchangeWeak(&${ptr}, oldValue, newValue);
+                  if res.exchanged {
+                    break;
+                  }
+                  oldValue = res.old_value;
+                }
+              }`;
+
+  switch (reduction) {
+    case 'none':
+      return `${ptr}=${v};`;
+    case 'add':
+      if (type === 'i32') {
+        return `atomicAdd(&${ptr}, bitcast<${type}>(${v}));`;
+      } else {
+        // atomicAdd only supports uint/int type. For float, we use
+        // atomicCompareExchangeWeak to simulate.
+        return `
+              ${floatStart}bitcast<${type}>(oldValue) + (${v})${floatEnd}`;
+      }
+    case 'max':
+      if (type === 'i32') {
+        return `atomicMax(&${ptr}, bitcast<${type}>(${v}));`;
+      } else {
+        // atomicMax only supports uint/int type. For float, we use
+        // atomicCompareExchangeWeak to simulate.
+        return `
+                ${floatStart}max(bitcast<f32>(oldValue), (${v}))${floatEnd}`;
+      }
+    case 'min':
+      if (type === 'i32') {
+        return `atomicMin(&${ptr}, bitcast<${type}>(${v}));`;
+      } else {
+        // atomicMin only supports uint/int type. For float, we use
+        // atomicCompareExchangeWeak to simulate.
+        return `${floatStart}min(bitcast<${type}>(oldValue), (${v}))${floatEnd}`;
+      }
+    case 'mul':
+      // atomicMul is not supported, we use atomicCompareExchangeWeak to simulate.
+      return `${floatStart}(bitcast<${type}>(oldValue) * (${v}))${floatEnd}`;
+
+    default:
+      throw new Error(`Reduction ${reduction} is not supported.`);
+  }
+};
+
+const createScatterNDProgramInfo = (inputs: readonly TensorView[], attributes: ScatterNDAttributes): ProgramInfo => {
+  const inputShape = inputs[0].dims;
+  const indicesShape = inputs[1].dims;
+  const outputShape = inputShape;
+  // TODO: support bool with componenets 4.
+  const components = 1;
+  const outputSize = Math.ceil(ShapeUtil.size(indicesShape) / components);
+  const lastIndexDimension = indicesShape[indicesShape.length - 1];
+  const numUpdatesElements = ShapeUtil.sizeFromDimension(inputShape, lastIndexDimension);
+
+  const programUniforms: ProgramUniform[] = [
+    { type: DataType.uint32, data: outputSize },
+    { type: DataType.uint32, data: lastIndexDimension },
+    { type: DataType.uint32, data: numUpdatesElements },
+    ...createTensorShapeVariables(inputs[1].dims, inputs[2].dims, outputShape),
+  ];
+
+  const getShaderSource = (shaderHelper: ShaderHelper) => {
+    const indices = inputVariable('indices', inputs[1].dataType, inputs[1].dims.length);
+    const updates = inputVariable('updates', inputs[2].dataType, inputs[2].dims.length, components);
+    const output =
+      attributes.reduction !== 'none' && attributes.reduction !== ''
+        ? atomicOutputVariable('output', inputs[0].dataType, outputShape.length)
+        : outputVariable('output', inputs[0].dataType, outputShape.length, components);
```

Comment:
It seems that if you go to the atomic path, only 'u32|i32|f32' are supported. Otherwise, JsepSupportedDataTypes are supported, which also include the f16. 

But f16 is a very important type. It seems that it's not supported if attributes.reduction is not none. I would suggest you also support it by another algorithm if atomic is not suitable for it.

Reply:
An idea to process f16 in my mind:
Convert f16 to f32, then use the atomic op. After the scatternd process, do a cast from f32 to f16 for f16 type. And return the casted result as the final ScatterND's result.

Reply:
Jiajia, good to merge?

Reply:
It's ok to merge now. Let's support f16 in further iteration PRs.